### ISSUE TITLE
keys/LinearIndices for for non-Int indices

### ISSUE THREAD

**Issue by sostock** [JuliaLang/julia#32566]
Due to `LinearIndices` only supporting `Int`s, `keys` and `LinearIndices` do not work when the indices are not representable as `Int`s:

```julia
julia> keys(1:typemax(Int128))
ERROR: InexactError: trunc(Int64, 170141183460469231731687303715884105727)
[...]

julia> LinearIndices(1:typemax(Int128))
ERROR: InexactError: trunc(Int64, 170141183460469231731687303715884105727)
[...]
```

Related: #27090

**Comment by mbauman** [comment]
Yeah, this is a bit of a wart.  I have [an old branch](https://github.com/JuliaLang/julia/compare/master...mbauman:mb/parameterizedindices) where I started parameterizing all these indices structures, but I recall hitting a number of roadblocks making this trickier than it first appears.

In some senses, I'm ok with `LinearIndices` being restricted to only being integer values — you're explicitly requesting a "memory offset"-like representation of the array, so an inability to represent values larger than `Int` is reasonable.  The trouble is that `keys` should definitely always work, and right now there's a nice symmetry that it always returns either a `Linear-` or `CartesianIndices`.

The easy fix is to just patch up `keys(::AbstractRange)` to deal with these cases, but it'd be nice to find a generic behavior that's a bit more friendly to non-Int indices.
